dvc/compare.py

```diff
@@ -0,0 +1,298 @@
+
+    def __getitem__(self, item: Union[int, slice]):
+        func = itemgetter(item)
+        it = map(func, self.columns)
+        if isinstance(item, slice):
+            it = map(list, zip(*it))
+        return list(it)
+
+    @overload
+    def __setitem__(self, item: int, value: Sequence["CellT"]) -> None:
+        ...
+
+    @overload
+    def __setitem__(
+        self, item: slice, value: Iterable[Sequence["CellT"]]
+    ) -> None:
+        ...
+
+    def __setitem__(self, item, value) -> None:
+        it = value
+        if isinstance(item, slice):
+            n = len(self.columns)
+            normalized_rows = (
+                chain(val, repeat(self._fill_value, n - len(val)))
+                for val in value
+            )
+            # we need to transpose those rows into columnar format
+            # as we work in terms of column-based arrays
+            it = zip(*normalized_rows)
+
+        for i, col in self._iter_col_row(it):
+            col[item] = i
+
+    def __delitem__(self, item: Union[int, slice]) -> None:
+        for col in self.columns:
+            del col[item]
+
+    def __len__(self) -> int:
+        return len(self.columns[0])
+
+    @property
+    def shape(self) -> Tuple[int, int]:
+        return len(self.columns), len(self)
+
+    def drop(self, *col_names: str) -> None:
+        for col_name in col_names:
+            self._keys.remove(col_name)
+            self._columns.pop(col_name)
+
+    def rename(self, from_col_name: str, to_col_name: str) -> None:
+        self._columns[to_col_name] = self._columns.pop(from_col_name)
+        self._keys[self._keys.index(from_col_name)] = to_col_name
+
+    def project(self, *col_names: str) -> None:
+        self.drop(*(set(self._keys) - set(col_names)))
+        self._keys = list(col_names)
+
+    def to_csv(self) -> str:
+        import csv
+        from io import StringIO
+
+        buff = StringIO()
+        writer = csv.writer(buff)
+        writer.writerow(self.keys())
+
+        for row in self:
+            writer.writerow(row)
+        return buff.getvalue()
+
+    def add_column(self, name: str) -> None:
+        self._columns[name] = Column([self._fill_value] * len(self))
+        self._keys.append(name)
+
+    def row_from_dict(self, d: Dict[str, "CellT"]) -> None:
+        keys = self.keys()
+        for key in d:
+            if key not in keys:
+                self.add_column(key)
+
+        row: List["CellT"] = [
+            with_value(d.get(key), self._fill_value) for key in self.keys()
+        ]
+        self.append(row)
+
+    def render(self, **kwargs: Any):
+        from dvc.ui import ui
+
+        ui.table(self, headers=self.keys(), **kwargs)
+
+    def as_dict(self):
+        keys = self.keys()
+        return [dict(zip(keys, row)) for row in self]
+
+
+def _format_field(val: Any, precision: int = None) -> str:
+    def _format(_val):
+        if isinstance(_val, float) and precision:
+            fmt = f"{{:.{precision}f}}"
```

Comment:
I changed the precision from `g` specifier to `f` now, which is how other `params/metrics` diff worked.
This might affect experiments.

Reply:
This will break the experiments table, for very big or very small numbers we want to display them in scientific notation (with the correct precision), and for everything else we want to display them as regular decimals, which is why it was `g` in experiments. It seems like this should also just be the default behavior everywhere in DVC.

Reply:
according to the python docs, to keep the number of digits when it's displayed as a regular decimal, you can use `#` in conjunction with `g`

so what we probably want here is 
`f"{{:#.{precision}g}}"`

Reply:
`#` specifier only affects the trailing zeros. The problem with `g` specifier is that it uses ` P − (exp + 1)` precision for fixed-point formatting and `P-1` for exponent-based formatting.

I guess it depends on what we are aiming for with `precision`, precision in terms of significant digits or in terms of fractional digits. The docs [points to the latter](https://dvc.org/doc/command-reference/exp/show#options):

>  - round decimal values to n digits of precision (5 by default). Applies to metrics only.

Reply:
I think we should go with precision meaning significant digits and update the docs accordingly. @dberenbaum thoughts on this?

The switch to `g` for exp commands was made specifically because users had metrics/params with very small magnitudes, and using `f` will result in rounding the number down to zero, which is definitely not what DVC should be displaying in the exp table (or in the other diff/show commands).

Reply:
@pmrowla, significant digits makes sense to me as well. I changed to `g`, but I can rollback later if we decide so (changing to `g` does affect metrics/params diff/show). I'll update docs separately.